File: TaylorCode/app/src/main/java/test/taylor/com/taylorcode/ui/window/FloatWindow.java
package test.taylor.com.taylorcode.ui.window;

import android.animation.ValueAnimator;
import android.content.Context;
import android.graphics.PixelFormat;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.Display;
import android.view.GestureDetector;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.view.WindowManager;
import android.view.animation.LinearInterpolator;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

/**
 * suspending window in app,it shows throughout the whole app
 */
public class FloatWindow implements View.OnTouchListener {

    private static final int DEFAULT_WIDTH = 100;
    private static final int DEFAULT_HEIGHT = 100;
    private static final long WELT_ANIMATION_DURATION = 150;
    /**
     * the current showing content view of window
     */
//    private View windowView;
    /**
     * the current showing layout param for windowView
     */
//    private WindowManager.LayoutParams layoutParam;
    /**
     * several window content stored by String tag ;
     */
    private HashMap<String, WindowContent> windowContentMap;
    private WindowContent windowContent;
    private int width;
    private int height;
    //    private Rect windowRect;
    private int lastTouchX;
    private int lastTouchY;
    private int lastWeltX;
    private int screenWidth;
    private int screenHeight;
    private Context context;
    private GestureDetector gestureDetector;
    private OnWindowViewClickListener clickListener;
    private OnWindowStatusChangeListener onWindowStatusChangeListener;
    /**
     * this list records the activities which shows this window
     */
    private List<Class> whiteList;
    /**
     * if true,whiteList will be used to depend which activity could show window
     * if false,all activities in app is allow to show window
     */
    private boolean enableWhileList;
    /**
     * whether to show or hide this window
     * true by default
     */
//    private boolean enable = true;

    private static volatile FloatWindow INSTANCE;
    /**
     * the animation make window stick to the left or right side of screen
     */
    private ValueAnimator weltAnimator;

    public static FloatWindow getInstance() {
        if (INSTANCE == null) {
            synchronized (FloatWindow.class) {
                if (INSTANCE == null) {
                    //in case of memory leak for singleton
                    INSTANCE = new FloatWindow();
                }
            }
        }
        return INSTANCE;
    }

    public FloatWindow init(Context context) {
        whiteList = new ArrayList<>();
        WindowManager windowManager = ((WindowManager) context.getSystemService(Context.WINDOW_SERVICE));
        prepareScreenDimension(windowManager);
        gestureDetector = new GestureDetector(context, new GestureListener());
        return this;
    }


    private int getNavigationBarHeight(Context context) {
        int result = 0;
        int resourceId = 0;
        int rid = context.getResources().getIdentifier("config_showNavigationBar", "bool", "android");
        if (rid != 0) {
            resourceId = context.getResources().getIdentifier("navigation_bar_height", "dimen", "android");
            return context.getResources().getDimensionPixelSize(resourceId);
        } else {
            return 0;
        }
    }

    private FloatWindow() {
    }

    public void updateWindowView(IWindowUpdater updater) {
        if (updater != null) {
            updater.updateWindowView(windowContent.windowView);
        }
        WindowManager windowManager = ((WindowManager) context.getSystemService(Context.WINDOW_SERVICE));
        if (windowManager != null&&windowContent != null && windowContent.windowView != null && windowContent.windowView.getParent() != null){
            windowManager.updateViewLayout(windowContent.windowView, windowContent.layoutParams);
        }
    }


    public void setOnClickListener(OnWindowViewClickListener listener) {
        clickListener = listener;
    }

    public void setWidth(int width) {
        this.width = width;
    }

    public void setHeight(int height) {
        this.height = height;
    }

    public void setEnable(boolean enable, String tag) {
        if (windowContentMap == null) {
            return;
        }
        WindowContent windowContent = windowContentMap.get(tag);
        if (windowContent == null) {
            throw new RuntimeException("no such window view,please invoke setView() first");
        }
        windowContent.enable = enable;
    }

    public void setOnWindowStatusChangeListener(OnWindowStatusChangeListener onWindowStatusChangeListener) {
        this.onWindowStatusChangeListener = onWindowStatusChangeListener;
    }

    public void show(Context context, String tag) {
        if (windowContentMap == null) {
            return;
        }
        windowContent = windowContentMap.get(tag);
        if (windowContent == null) {
            Log.v("ttaylor", "there is no view to show,please invoke setView() first");
            return;
        }
        if (!windowContent.enable) {
            return;
        }
        if (windowContent.windowView == null) {
            return;
        }
        this.context = context;
        windowContent.layoutParams = generateLayoutParam(screenWidth, screenHeight);
        //in case of "IllegalStateException :has already been added to the window manager."
        if (windowContent.windowView.getParent() == null) {
            WindowManager windowManager = ((WindowManager) this.context.getSystemService(Context.WINDOW_SERVICE));
            windowManager.addView(windowContent.windowView, windowContent.layoutParams);
            if (onWindowStatusChangeListener != null) {
                onWindowStatusChangeListener.onShow();
            }
        }
    }

    private WindowManager.LayoutParams generateLayoutParam(int screenWidth, int screenHeight) {
        if (context == null) {
            return new WindowManager.LayoutParams();
        }

        WindowManager.LayoutParams layoutParams = new WindowManager.LayoutParams();
        layoutParams.type = WindowManager.LayoutParams.TYPE_APPLICATION;
        layoutParams.format = PixelFormat.TRANSLUCENT;
        layoutParams.flags = WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE | WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS;
        layoutParams.gravity = Gravity.LEFT | Gravity.TOP;
        layoutParams.width = this.width == 0 ? DEFAULT_WIDTH : this.width;
        layoutParams.height = this.height == 0 ? DEFAULT_HEIGHT : this.height;
        int defaultX = screenWidth - layoutParams.width;
        int defaultY = 2 * screenHeight / 3;
        layoutParams.x = windowContent.layoutParams == null ? defaultX : windowContent.layoutParams.x;
        layoutParams.y = windowContent.layoutParams == null ? defaultY : windowContent.layoutParams.y;
        return layoutParams;
    }

    public WindowManager.LayoutParams getLayoutParam() {
        if (windowContent != null) {
            return windowContent.layoutParams;
        }
        return null;
    }

    public void dismiss() {
        WindowManager windowManager = ((WindowManager) context.getSystemService(Context.WINDOW_SERVICE));
        if (windowManager != null) {
            //in case of "IllegalStateException :not attached to window manager."
            if (windowContent != null && windowContent.windowView != null && windowContent.windowView.getParent() != null) {
                windowManager.removeViewImmediate(windowContent.windowView);
                if (onWindowStatusChangeListener != null) {
                    onWindowStatusChangeListener.onDismiss();
                }
            }
        }
    }

    public boolean isShowing() {
        if (windowContent != null) {
            return false;
        }
        if (windowContent.windowView == null) {
            return false;
        }
        if (windowContent.windowView.getParent() == null) {
            return false;
        } else {
            return true;
        }
    }

    public void setWhiteList(List<Class> whiteList) {
        enableWhileList = true;
        this.whiteList = whiteList;
    }

    /**
     * invoking this method is a must ,or window has no content to show
     *
     * @param view the content view of window
     * @param tag
     * @return
     */
    public FloatWindow setView(View view, String tag) {
        if (windowContentMap == null) {
            windowContentMap = new HashMap<>();
        }
        WindowContent windowContent = new WindowContent();
        windowContent.windowView = view;
        windowContentMap.put(tag, windowContent);
        windowContent.windowView.setOnTouchListener(this);
        return this;
    }

    private void saveWindowContent(String tag) {
        if (windowContentMap == null) {
            windowContentMap = new HashMap<>();
        }
        windowContentMap.put(tag, windowContent);
    }

    @Override
    public boolean onTouch(View v, MotionEvent event) {
//        int action = event.getAction();
//
//        switch (action) {
//            case MotionEvent.ACTION_DOWN:
//                onActionDown(event);
//                break;
//            case MotionEvent.ACTION_MOVE:
//                onActionMove(event);
//                break;
//            default:
//                break;
//        }
        //let GestureDetector take care of touch event,in order to parsing touch event into different gesture
        gestureDetector.onTouchEvent(event);
        //there is no ACTION_UP event in GestureDetector
        int action = event.getAction();
        switch (action) {
            case MotionEvent.ACTION_UP:
                onActionUp(event, screenWidth, width);
                break;
            default:
                break;
        }
        return true;
    }

    private void onActionUp(MotionEvent event, int screenWidth, int width) {
        int upX = ((int) event.getRawX());
        int endX;
        if (upX > screenWidth / 2) {
            endX = screenWidth - width;
        } else {
            endX = 0;
        }

        if (weltAnimator == null) {
            weltAnimator = ValueAnimator.ofInt(windowContent.layoutParams.x, endX);
            weltAnimator.setInterpolator(new LinearInterpolator());
            weltAnimator.setDuration(WELT_ANIMATION_DURATION);
            weltAnimator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
                @Override
                public void onAnimationUpdate(ValueAnimator animation) {
                    int x = ((int) animation.getAnimatedValue());
                    if (windowContent.layoutParams != null) {
                        windowContent.layoutParams.x = x;
                    }
                    WindowManager windowManager = ((WindowManager) context.getSystemService(Context.WINDOW_SERVICE));
                    if (windowManager != null && windowContent.windowView.getParent() != null) {
                        windowManager.updateViewLayout(windowContent.windowView, windowContent.layoutParams);
                    }
                }
            });
        }
        weltAnimator.setIntValues(windowContent.layoutParams.x, endX);
        weltAnimator.start();
    }


    private void onActionMove(MotionEvent event) {
        int currentX = (int) event.getRawX();
        int currentY = (int) event.getRawY();
        int dx = currentX - lastTouchX;
        int dy = currentY - lastTouchY;

        windowContent.layoutParams.x += dx;
        windowContent.layoutParams.y += dy;
//        windowRect.set(layoutParam.x, layoutParam.y, layoutParam.x + layoutParam.width, layoutParam.y + layoutParam.height);
        WindowManager windowManager = ((WindowManager) context.getSystemService(Context.WINDOW_SERVICE));
        if (windowManager != null) {
            int rightMost = screenWidth - windowContent.layoutParams.width;
            int leftMost = 0;
            int topMost = 0;
            int bottomMost = screenHeight - windowContent.layoutParams.height - getNavigationBarHeight(context);
            WindowManager.LayoutParams partnerParam = null;
            if (onWindowStatusChangeListener != null) {
                partnerParam = onWindowStatusChangeListener.onWindowMove(dx, dy, screenWidth, screenHeight, windowContent.layoutParams);
            }
            //adjust move area according to partner window
            if (partnerParam != null) {
                if (partnerParam.x < windowContent.layoutParams.x) {
                    leftMost = partnerParam.width - windowContent.layoutParams.width / 2;
                } else if (partnerParam.x > windowContent.layoutParams.x) {
                    rightMost = screenWidth - (windowContent.layoutParams.width / 2 + partnerParam.width);
                }
            }

            //make window float inside screen
            if (windowContent.layoutParams.x < leftMost) {
                windowContent.layoutParams.x = leftMost;
            }
            if (windowContent.layoutParams.x > rightMost) {
                windowContent.layoutParams.x = rightMost;
            }
            if (windowContent.layoutParams.y < topMost) {
                windowContent.layoutParams.y = topMost;
            }
            if (windowContent.layoutParams.y > bottomMost) {
                windowContent.layoutParams.y = bottomMost;
            }
            windowManager.updateViewLayout(windowContent.windowView, windowContent.layoutParams);
        }
        lastTouchX = currentX;
        lastTouchY = currentY;
    }

    private void onActionDown(MotionEvent event) {
        lastTouchX = (int) event.getRawX();
        lastTouchY = (int) event.getRawY();
    }


    private void prepareScreenDimension(WindowManager windowManager) {
        if (screenWidth != 0 && screenHeight != 0) {
            return;
        }
        if (windowManager != null) {
            DisplayMetrics dm = new DisplayMetrics();
            Display display = windowManager.getDefaultDisplay();
            if (display != null) {
                windowManager.getDefaultDisplay().getMetrics(dm);
                screenWidth = dm.widthPixels;
                screenHeight = dm.heightPixels;
            }
        }
    }

    /**
     * let ui decide how to update window
     */
    public interface IWindowUpdater {
        void updateWindowView(View windowView);
    }

    public interface OnWindowViewClickListener {
        void onWindowViewClick();
    }

    public interface OnWindowStatusChangeListener {
        void onShow();

        void onDismiss();

        WindowManager.LayoutParams onWindowMove(float dx, float dy, int screenWidth, int screenHeight, WindowManager.LayoutParams layoutParams);
    }

    private class GestureListener implements GestureDetector.OnGestureListener {

        @Override
        public boolean onDown(MotionEvent e) {
            onActionDown(e);
            return false;
        }

        @Override
        public void onShowPress(MotionEvent e) {
        }

        @Override
        public boolean onSingleTapUp(MotionEvent e) {
            if (clickListener != null) {
                clickListener.onWindowViewClick();
            }
            return false;
        }

        @Override
        public boolean onScroll(MotionEvent e1, MotionEvent e2, float distanceX, float distanceY) {
            onActionMove(e2);
            return true;
        }

        @Override
        public void onLongPress(MotionEvent e) {
        }

        @Override
        public boolean onFling(MotionEvent e1, MotionEvent e2, float velocityX, float velocityY) {
            return false;
        }

    }

    public void onDestroy() {
        if (windowContentMap != null) {
            windowContentMap.clear();
        }
    }

    private class WindowContent {
        /**
         * the content view of window
         */
        public View windowView;
        /**
         * the layout param of window content view
         */
        public WindowManager.LayoutParams layoutParams;
        /**
         * whether this window content is allow to show
         */
        public boolean enable = true;

    }
}

File: TaylorCode/app/src/main/java/test/taylor/com/taylorcode/ui/window/FloatWindow.kt
package test.taylor.com.taylorcode.ui.window

import android.animation.ValueAnimator
import android.content.Context
import android.graphics.PixelFormat
import android.util.DisplayMetrics
import android.util.Log
import android.view.*
import android.view.animation.LinearInterpolator
import java.util.*

/**
 * a window shows above an activity.
 * it can be dragged by finger and if your finger leave the screen, it will stick to the nearest border of the screen
 * it can be clicked
 */
object FloatWindow : View.OnTouchListener {
    /**
     * several window content stored by String tag ;
     */
    private var windowInfoMap: HashMap<String, WindowInfo?> = HashMap()
    private var windowInfo: WindowInfo? = null
    private var lastTouchX: Int = 0
    private var lastTouchY: Int = 0
    private val lastWeltX: Int = 0
    private var screenWidth: Int = 0
    private var screenHeight: Int = 0
    private var context: Context? = null
    private var gestureDetector: GestureDetector = GestureDetector(context, GestureListener())
    private var clickListener: WindowClickListener? = null
    private var windowStateListener: WindowStateListener? = null
    private var dragEnable: Boolean = false
    /**
     * this list records the activities which shows this window
     */
    private var whiteList: List<Class<Any>>? = mutableListOf()
    /**
     * if true,whiteList will be used to depend which activity could show window
     * if false,all activities in app is allow to show window
     */
    private var enableWhileList: Boolean = false
    /**
     * the animation make window stick to the left or right side of screen
     */
    private var weltAnimator: ValueAnimator? = null

    val layoutParam: WindowManager.LayoutParams?
        get() = windowInfo?.layoutParams

    val isShowing: Boolean
        get() = windowInfo?.view?.parent != null

    /**
     * listener invoked when touch outside of [FloatWindow]
     */
    private var onTouchOutside: (() -> Unit)? = null


    private fun getNavigationBarHeight(context: Context?): Int {
        val rid = context?.resources?.getIdentifier("config_showNavigationBar", "bool", "android")
        return if (rid != 0) {
            context?.resources?.getIdentifier("navigation_bar_height", "dimen", "android")?.let {
                context.resources?.getDimensionPixelSize(it)
            } ?: 0
        } else {
            0
        }
    }

    fun updateWindowView(updater: IWindowUpdater?) {
        updater?.updateWindowView(windowInfo?.view)
        val windowManager = context?.getSystemService(Context.WINDOW_SERVICE) as WindowManager
        if (windowInfo?.hasParent().value()) {
            windowManager.updateViewLayout(windowInfo?.view, windowInfo?.layoutParams)
        }
    }

    fun setClickListener(listener: WindowClickListener) {
        clickListener = listener
    }

    fun setEnable(enable: Boolean, tag: String) {
        val windowInfo = windowInfoMap[tag]
            ?: throw RuntimeException("no such window view,please invoke setView() first")
        windowInfo.enable = enable
    }

    fun setWindowStateListener(windowStateListener: WindowStateListener) {
        this.windowStateListener = windowStateListener
    }

    /**
     * show float window, every window has a tag
     * @param tag a unique tag for a window, if showing the previous window without providing [windowInfo], we will looking for it in [windowInfoMap]
     * @param windowInfo the necessary information for showing float window, it will be kept in [windowInfoMap] with the key [tag]
     * @param x the horizontal position of float window according to the left top of screen
     * @param y the vertical position of float window according to the left top of screen
     */
    fun show(
        context: Context,
        tag: String,
        windowInfo: WindowInfo? = windowInfoMap[tag],
        x: Int = windowInfo?.layoutParams?.x.value(),
        y: Int = windowInfo?.layoutParams?.y.value(),
        dragEnable: Boolean
    ) {
        if (windowInfo == null) {
            Log.v("ttaylor", "there is no view to show,please creating the right WindowInfo object")
            return
        }
        if (!windowInfo.enable) {
            return
        }
        if (windowInfo.view == null) {
            return
        }
        this.windowInfo = windowInfo
        this.dragEnable = dragEnable
        windowInfoMap[tag] = windowInfo
        windowInfo.view?.setOnTouchListener(this)
        this.context = context
        windowInfo.layoutParams = createLayoutParam(x, y)
        //in case of "IllegalStateException :has already been added to the window manager."
        if (!windowInfo.hasParent().value()) {
            val windowManager =
                this.context?.getSystemService(Context.WINDOW_SERVICE) as WindowManager
            prepareScreenDimension(windowManager)
            windowManager.addView(windowInfo.view, windowInfo.layoutParams)
            windowStateListener?.onWindowShow()
        }
    }

    private fun createLayoutParam(x: Int, y: Int): WindowManager.LayoutParams {
        if (context == null) {
            return WindowManager.LayoutParams()
        }

        return WindowManager.LayoutParams().apply {
            type = WindowManager.LayoutParams.TYPE_APPLICATION
            format = PixelFormat.TRANSLUCENT
            flags =
                WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE or WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS
            gravity = Gravity.START or Gravity.TOP
            width = windowInfo?.width.value()
            height = windowInfo?.height.value()
            this.x = x
            this.y = y
        }
    }

    fun dismiss() {
        val windowManager = context?.getSystemService(Context.WINDOW_SERVICE) as WindowManager
        //in case of "IllegalStateException :not attached to window manager."
        if (windowInfo?.hasParent().value()) {
            windowManager.removeViewImmediate(windowInfo?.view)
            windowStateListener?.onWindowDismiss()
        }
    }

    fun setWhiteList(whiteList: List<Class<Any>>) {
        enableWhileList = true
        this.whiteList = whiteList
    }


    override fun onTouch(v: View, event: MotionEvent): Boolean {
        //handle outside touch event
        val x = event.x.toInt()
        val y = event.y.toInt()

        if (event.action == MotionEvent.ACTION_OUTSIDE) {
            onTouchOutside?.invoke()
            return true
        }

        //let GestureDetector take care of touch event,in order to parsing touch event into different gesture
        gestureDetector.onTouchEvent(event).takeIf { !it }?.also {
            //there is no ACTION_UP event in GestureDetector
            val action = event.action
            when (action) {
                MotionEvent.ACTION_UP -> onActionUp(event, screenWidth, windowInfo?.width ?: 0)
                else -> {
                }
            }
        }
        return true
    }

    private fun isTouchOutside(x: Int, y: Int) =
        x < 0 || x >= windowInfo?.view?.width ?: Int.MAX_VALUE || y < 0 || y >= windowInfo?.view?.height ?: Int.MAX_VALUE

    /**
     * set whether outside touch event could be detected
     */
    fun setOutsideTouchable(enable: Boolean, onTouchOutside: (() -> Unit)? = null) {
        if (enable) windowInfo?.layoutParams?.let { layoutParams ->
            layoutParams.flags =
                layoutParams.flags or WindowManager.LayoutParams.FLAG_WATCH_OUTSIDE_TOUCH
            this.onTouchOutside = onTouchOutside
        }
    }

    private fun onActionUp(event: MotionEvent, screenWidth: Int, width: Int) {
        if (!windowInfo?.hasView().value()) {
            return
        }
        val upX = event.rawX.toInt()
        val endX = if (upX > screenWidth / 2) {
            screenWidth - width
        } else {
            0
        }

        if (weltAnimator == null) {
            weltAnimator = ValueAnimator.ofInt(windowInfo?.layoutParams!!.x, endX).apply {
                interpolator = LinearInterpolator()
                duration = WELT_ANIMATION_DURATION
                addUpdateListener { animation ->
                    val x = animation.animatedValue as Int
                    if (windowInfo?.layoutParams != null) {
                        windowInfo?.layoutParams!!.x = x
                    }
                    val windowManager =
                        context?.getSystemService(Context.WINDOW_SERVICE) as WindowManager
                    if (windowInfo?.hasParent().value()) {
                        windowManager.updateViewLayout(windowInfo?.view, windowInfo?.layoutParams)
                    }
                }
            }
        }
        weltAnimator?.setIntValues(windowInfo?.layoutParams!!.x, endX)
        weltAnimator?.start()
    }


    private fun onActionMove(event: MotionEvent) {
        val currentX = event.rawX.toInt()
        val currentY = event.rawY.toInt()
        val dx = currentX - lastTouchX
        val dy = currentY - lastTouchY

        windowInfo?.layoutParams!!.x += dx
        windowInfo?.layoutParams!!.y += dy
        val windowManager = context?.getSystemService(Context.WINDOW_SERVICE) as WindowManager
        var rightMost = screenWidth - windowInfo?.layoutParams!!.width
        var leftMost = 0
        val topMost = 0
        val bottomMost =
            screenHeight - windowInfo?.layoutParams!!.height - getNavigationBarHeight(context)
        var partnerParam: WindowManager.LayoutParams? = null
        partnerParam = windowStateListener?.onWindowMove(
            dx.toFloat(),
            dy.toFloat(),
            screenWidth,
            screenHeight,
            windowInfo?.layoutParams
        )
        //adjust move area according to partner window
        if (partnerParam != null) {
            if (partnerParam.x < windowInfo?.layoutParams!!.x) {
                leftMost = partnerParam.width - windowInfo?.layoutParams!!.width / 2
            } else if (partnerParam.x > windowInfo?.layoutParams!!.x) {
                rightMost =
                    screenWidth - (windowInfo?.layoutParams!!.width / 2 + partnerParam.width)
            }
        }

        //make window float inside screen
        if (windowInfo?.layoutParams!!.x < leftMost) {
            windowInfo?.layoutParams!!.x = leftMost
        }
        if (windowInfo?.layoutParams!!.x > rightMost) {
            windowInfo?.layoutParams!!.x = rightMost
        }
        if (windowInfo?.layoutParams!!.y < topMost) {
            windowInfo?.layoutParams!!.y = topMost
        }
        if (windowInfo?.layoutParams!!.y > bottomMost) {
            windowInfo?.layoutParams!!.y = bottomMost
        }
        windowManager.updateViewLayout(windowInfo?.view, windowInfo?.layoutParams)
        lastTouchX = currentX
        lastTouchY = currentY
    }

    private fun onActionDown(event: MotionEvent) {
        lastTouchX = event.rawX.toInt()
        lastTouchY = event.rawY.toInt()
    }

    private fun prepareScreenDimension(windowManager: WindowManager?) {
        if (screenWidth != 0 && screenHeight != 0) {
            return
        }
        if (windowManager != null) {
            val dm = DisplayMetrics()
            val display = windowManager.defaultDisplay
            if (display != null) {
                windowManager.defaultDisplay.getMetrics(dm)
                screenWidth = dm.widthPixels
                screenHeight = dm.heightPixels
            }
        }
    }

    /**
     * let ui decide how to update window
     */
    interface IWindowUpdater {
        fun updateWindowView(windowView: View?)
    }

    /**
     * let ui decide what to do after window clicked
     */
    interface WindowClickListener {
        fun onWindowClick(windowInfo: WindowInfo?): Boolean
    }

    interface WindowStateListener {
        fun onWindowShow()

        fun onWindowDismiss()

        fun onWindowMove(
            dx: Float,
            dy: Float,
            screenWidth: Int,
            screenHeight: Int,
            layoutParams: WindowManager.LayoutParams?
        ): WindowManager.LayoutParams
    }

    private class GestureListener : GestureDetector.OnGestureListener {

        override fun onDown(e: MotionEvent): Boolean {
            onActionDown(e)
            return false
        }

        override fun onShowPress(e: MotionEvent) {}

        override fun onSingleTapUp(e: MotionEvent): Boolean {
            return clickListener?.onWindowClick(windowInfo) ?: false
        }

        override fun onScroll(
            e1: MotionEvent,
            e2: MotionEvent,
            distanceX: Float,
            distanceY: Float
        ): Boolean {
            if (dragEnable) {
                onActionMove(e2)
                return true
            }
            return false
        }

        override fun onLongPress(e: MotionEvent) {}

        override fun onFling(
            e1: MotionEvent,
            e2: MotionEvent,
            velocityX: Float,
            velocityY: Float
        ): Boolean {
            return false
        }

    }

    fun onDestroy() {
        windowInfoMap.clear()
        context = null
    }

    class WindowInfo(var view: View?) {
        /**
         * the layout param of window content view
         */
        var layoutParams: WindowManager.LayoutParams? = null
        /**
         * whether this window content is allow to show
         */
        var enable = true
        /**
         * the width of window content
         */
        var width: Int = 0
        /**
         * the height of window content
         */
        var height: Int = 0

        fun hasView() = view != null && layoutParams != null

        fun hasParent() = hasView() && view?.parent != null

    }

    private val WELT_ANIMATION_DURATION: Long = 150

    fun Boolean?.value() = this ?: false
    fun Int?.value() = this ?: 0
}
